### ISSUE TITLE
Possible SQLite issue with replicating and saving at the same time

### ISSUE THREAD

**Issue by borrrden** [couchbase/couchbase-lite-core#154]
Original issue reference: https://github.com/couchbase/couchbase-lite-net/issues/872

Reproduced on .NET Core with both macOS and Windows.  This could be a .NET implementation bug but I'm not sure what could be going wrong in this case.  It seems like the replicator's enumeration is causing the save operations to return SQLITE_BUSY.  Here is a verbose log of the run:

https://gist.github.com/borrrden/532c84e2ab791e23a19b29dbe6e0896a

To summarize the repro process:  

1. Start a continuous push and pull (or just push) replication
2. Add a number of documents (in the case of the above, 1000)
3. Start updating them sequentially at 300 ms intervals (the first update causes the locked issue)

Adding a 200 ms delay between 2 and 3 causes the issue to go away.

**Comment by borrrden** [comment]
Note: the test thread changes from 8 to 15 in the middle due to awaiting async stuff.  I redid the test without async stuff in the test and the thread remained 8 with the same results.

**Comment by snej** [comment]
It would help if you could find out where the error is originating from. Maybe add a line of code to the `SQLite::Exception` constructors, to log and dump a backtrace?

**Comment by borrrden** [comment]
Here is a stack trace from where the error is thrown.

> at SQLite::printStackTrace in d:\development\couchbase-lite-net-2\vendor\couchbase-lite-core\vendor\sqlitecpp\src\statement.cpp: line: 260: address: 0xE98E44A0
        at SQLite::Statement::exec in d:\development\couchbase-lite-net-2\vendor\couchbase-lite-core\vendor\sqlitecpp\src\statement.cpp: line: 333: address: 0xE98E3B50
        at litecore::SQLiteKeyStore::set in d:\development\couchbase-lite-net-2\vendor\couchbase-lite-core\litecore\storage\sqlitekeystore.cc: line: 263: address: 0xE98EB040
        at litecore::VersionedDocument::save in d:\development\couchbase-lite-net-2\vendor\couchbase-lite-core\litecore\revtrees\versioneddocument.cc: line: 88: address: 0xE98A7AD0
        at c4Internal::TreeDocument::save in d:\development\couchbase-lite-net-2\vendor\couchbase-lite-core\litecore\database\treedocument.cc: line: 210: address: 0xE986CE80
        at c4Internal::TreeDocument::putNewRevision in d:\development\couchbase-lite-net-2\vendor\couchbase-lite-core\litecore\database\treedocument.cc: line: 350: address: 0xE986CB90
        at c4doc_update in d:\development\couchbase-lite-net-2\vendor\couchbase-lite-core\c\c4document.cc: line: 367: address: 0xE9865920

**Comment by snej** [comment]
Weird. I don't understand why the busy-retry timeout isn't taking effect; I'll look at the code in the SQLite C++ wrapper to make sure it sets the timeout correctly.

But also I don't know why it would be encountering a busy database at all, since the DataFile::Shared class enforces that only a single instance can be in a transaction at once.

**Comment by snej** [comment]
The busy timeout is definitely set to 10sec. I don't think this is the problem. From the error message, the actual SQLite error seems to be SQLITE_LOCKED not SQLITE_BUSY. The [description of that error](https://www.sqlite.org/rescode.html#locked) says:

>The SQLITE_LOCKED result code differs from SQLITE_BUSY in that SQLITE_LOCKED indicates a conflict on the same database connection (or on a connection with a shared cache) whereas SQLITE_BUSY indicates a conflict with a different database connection, probably in a different process.

This gives me a strong suspicion that the problem might be unsafe multithreaded access to the database connection (`c4Database` or objects created from it), especially since the problem only manifests on one platform. For instance, I think that having an active `sqlite3_stmt` from a query enumeration can cause an `UPDATE` command to fail this way.

@borrrden, is there a way you could capture all of the native threads? Like, if you changed your `printStackTrace` call into something that would deliberately crash, generating a crash report that includes a thread dump? Because I'm guessing there's another thread that's doing something on the same db connection.

**Comment by borrrden** [comment]
The replicator is accessing it at the same time. I've followed the implementation pretty closely but maybe I've made a mistake somewhere. Should the replicator be copying the c4 database object when it gets created via c4repl_new or whatever?

**Comment by snej** [comment]
Yes, `c4repl_new` opens a new `c4Database`, so that shouldn't be a problem.

**Comment by borrrden** [comment]
The extended error code is 517 (SQLITE_BUSY_SNAPSHOT).  It seems that WAL mode has what looks to be a somewhat fragile setup.  

`sqlite3WalBeginWriteTransaction` is the only place that this error code is returned in the way that sqlite is compiled for us.  There is basically an increment counter for how many times the WAL has been written to and if the current thread doesn't match it, that's an error.

There is also only one place where this gets incremented, in `sqlite3WalFrames`.  This note in `sqlite3PagerBegin` also sheds some light;

```
/* Grab the write lock on the log file. If successful, upgrade to
      ** PAGER_RESERVED state. Otherwise, return an error code to the caller.
      ** The busy-handler is not invoked if another connection already
      ** holds the write-lock. If possible, the upper layer will call it.
      */
```

So it seems that the replicator and others are racing to get the write lock and the loser gets awarded with `SQLITE_BUSY`.  I guess that means we need to have retry logic?  Resetting the statement is not enough as it returns the same code that was generated during the step.

**Comment by borrrden** [comment]
Ok so it looks like the problem is a race between the time when the platform `SaveInto` method starts a transaction (via `InBatch()` in CBL) which boils down to "BEGIN" and the time when it actually makes its first write.  If anything changes in between those two periods then this happens and at the LiteCore level it is dubious to try to recover because the transactions are being controlled at the platforms level.  The current transaction needs to be finished, and another started.  

Another remedy would be to change the "BEGIN" to "BEGIN IMMEDIATE" which would block any writers past that point from writing until it is done.  This probably has performance implications so we might consider making an API for either situation (one that is known to write and one that is not).

**Comment by snej** [comment]
There shouldn't be multiple connections trying to write, because there is a lock (managed by DataFile::Shared) that allows only one DataFile object to begin a transaction on a database at a time. And no writes take place outside of transactions.

So it sounds like something is going wrong with this logic somehow. I don't know why it would only happen in CBL/.NET.

**Comment by borrrden** [comment]
c4db_getUUIDs is one of the things that came through on another thread, and it is not bound by the same protection (looks like a different mutex?).  That also lines up with another error that I saw sometimes regarding unhandled exception const char* (there is a line that says `throw "fail"` around there).

**Comment by borrrden** [comment]
Nevermind about the above....I keep going around in circles through the code.

**Comment by snej** [comment]
All write access goes through KeyStore, and KeyStore deliberately makes it impossible to call a database-modifying method without passing in a Transaction object.

Or that's the intention. Maybe there is a direct SQLite call inside SQLiteDataFile or SQLiteKeyStore that accidentally doesn't grab the transaction lock first? But why would it only show up on .NET?

**Comment by snej** [comment]
I'm embarrassed that those `throw "..."` statements are still there; those were placeholders. I just filed #159 to clean them up.

**Comment by borrrden** [comment]
There was only one that I noticed, and it was there.  I've got a good debug environment going so I can list off every thing that commits to the WAL.  I noticed that running the debugger sometimes slows things down enough to make things succeed on .NET, and actually only .NET Core is affected...UWP doesn't show this despite being from the same code base.  Although UWP has SQLite compiled for Windows Store vs desktop.  

However, I have a good debug setup going and I can list everything that commits to the WAL in the order that it does which I think will be helpful for further analysis.  I'll be on the lookout for violation of the mutexes.  I notice that every time this has happened the WAL counter has been 6 while the transaction that is ready to upgrade to write is on 5.  

As for why it only shows up on .NET...I have no idea but there are not a lot of places in this repro case for .NET to do something wrong.  All access in the entire test is from one thread (I changed it so that there is no more async).  It may not be that it only shows up on .NET, but that it only shows up on CMake compiles for LiteCore?
